include/oneapi/dpl/internal/reduce_by_segment_impl.h

```diff
@@ -138,6 +138,288 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+        oneapi::dpl::__ranges::__require_access(__cgh, __keys);
+        auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::write>(__cgh);
+
+        __cgh.parallel_for(sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+            auto __group = __item.get_group();
+            auto __group_id = __group.get_group_id();
+            int __local_id = __group.get_local_id();
+            int __global_id = __item.get_global_id();
+
+            int __start = __global_id * __vals_per_item;
+            int __end = sycl::minimum{}(__start + __vals_per_item, __n);
+            int __item_segments = 0;
+
+            // 1a. Work item scan to identify segment ends
+            for (int32_t __i = __start; __i < __end; ++__i)
+                if (__n - 1 == __i || __keys[__i] != __keys[__i + 1])
+                    ++__item_segments;
+
+            // 1b. Work group reduction
+            auto __num_segs = sycl::reduce_over_group(__group, __item_segments, sycl::plus<>());
+
+            // 1c. First work item writes segment count to global memory
+            if (__local_id == 0)
+                __seg_ends_acc[__group_id] = __num_segs;
+        });
+    });
+
+    // 2. Work group reduction
+    auto __wg_reduce = __exec.queue().submit([&](sycl::handler& __cgh) {
+        __cgh.depends_on(__seg_end_identification);
+        oneapi::dpl::__ranges::__require_access(__cgh, __keys, __out_keys, __out_values, __values);
+
+        auto __partials_acc = __partials.template get_access<sycl::access_mode::read_write>(__cgh);
+        auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::write>(__cgh);
+        auto __loc_acc = sycl::accessor<__val_type, 1, sycl::access::mode::read_write, sycl::access::target::local>{
+            2 * __wgroup_size, __cgh};
+
+        __cgh.parallel_for(sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+            ::std::array<__val_type, __vals_per_item> __loc_partials;
+
+            auto __group = __item.get_group();
+            auto __group_id = __group.get_group_id();
+            auto __local_id = __group.get_local_id();
+            auto __global_id = __item.get_global_id();
+
+            __val_type __accumulator{};
+
+            // 2a. Compute the number of segments in prior workgroups. Do this collectively in
+            // subgroups to eliminate barriers.
+            auto __start_ptr = __seg_ends_acc.get_pointer();
+            auto __end_ptr = __start_ptr + __group_id;
+
+            int32_t __wg_num_prior_segs =
+                sycl::joint_reduce(__item.get_sub_group(), __start_ptr, __end_ptr, sycl::plus<>());
+
+            // 2b. Perform a serial scan within the work item over assigned elements. Store partial
+            // reductions in work group local memory.
+            int32_t __start = __global_id * __vals_per_item;
```

Comment:
For result of __global_id * __vals_per_item you use int32_t here, in previous calculation - int